`PD_INSTR` & `PD_INSTR_DETECTOR`: what _is_ a detector in pdCIF?

rowlesmr opened the pull request:
From discussions on what a detector is. See:

I though I'd pull it out into its own branch so we can have a look at what it would look like

rowlesmr commented:
Low hanging fruit - need to add:
- `_pd_instr.radiation_id` - which source the instrument is attached to, not a key data name
- `_pd_instr_detector.instr_id` - which instrument the detector is attached to, not a key data name

1. What are the implications of defining a diffractogram as to the detector whence it came? that is, `_pd_diffractogram.detector_id` instead of `_pd_diffractogram.instr_id`
2. Are there any practical changes by asserting that `PD_DIFFRACTOGRAM`, `PD_DATA`, and `PD_INSTR_DETECTOR` are all 1D?
3. Does the 1D assertion and defining a diffractogram in terms of the detector it came from interact, in light of the existence of `_pd_meas.detector_id`? Note that there is currently no `_pd_proc.detector_id`.
4. How do we link a set of `_pd_proc.*` data names to the original raw data?
5. How do we link a `_pd_instr_detector.id` to the physical, potentially 1, 2, or 3D detector? should we?

jamesrhester commented:
> Questions to think about:
 
>  1. What are the implications of defining a diffractogram as to the detector whence it came? that is, `_pd_diffractogram.detector_id` instead of `_pd_diffractogram.instr_id`

As there is a direct link from `pd_instr_detector` to `pd_instr`, no information can be lost, so no consequences.
 
>  2. Are there any practical changes by defining `PD_DIFFRACTOGRAM`, `PD_DATA`, and `PD_INSTR_DETECTOR` to be 1D?

As there appears to be no consideration of 2D data in the powder dictionary, I think we are just making explicit what was previously implicit (as @rowlesmr notes above).

>  3. Does the 1D definition and defining a diffractogram in terms of the detector it came from interact, in light of the existence of `_pd_meas.detector_id`? Note that there is currently no `_pd_proc.detector_id`.

Very important question. Is a diffractogram the raw or processed data? Because if the former, there could be many detectors, but if the latter, then we are currently saying there is only one. There is a lot of untangling to do, because I've just realised that PD_DATA/PD_PROC/MEAS/CALC are all supposed to be aspects of one single category with identical key data names. I need to think about it.

>  4. How do we link a set of `_pd_proc.*` data names to the original raw data?

If `pd_meas` is missing, there is no raw 1D data so it is not possible to link to the raw data. If 2D data, then I've suggested a `_diffrn_detector.diffractogram_id`.

> 5. How do we link a `_pd_instr_detector.id` to the physical, potentially 1, 2, or 3D detector? should we?

I think this category should be limited to 1D detectors in the first instance. Although we could think about adding a link from `diffrn_detector` to `pd_instr_detector` if the data names make sense for an area detector (e.g. distances might need to be defined as along the normal to the detector and so on).

rowlesmr commented:
> > 2\. Are there any practical changes by defining `PD_DIFFRACTOGRAM`, `PD_DATA`, and `PD_INSTR_DETECTOR` to be 1D?
> 
> As there appears to be no consideration of 2D data in the powder dictionary, I think we are just making explicit what was previously implicit (as @rowlesmr notes above).

My thoughts as well.

rowlesmr commented:
> Does the 1D definition and defining a diffractogram in terms of the detector it came from interact, in light of the existence of _pd_meas.detector_id? Note that there is currently no _pd_proc.detector_id.\nVery important question. Is a diffractogram the raw or processed data? Because if the former, there could be many detectors, but if the latter, then we are currently saying there is only one. There is a lot of untangling to do, because I've just realised that PD_DATA/PD_PROC/MEAS/CALC are all supposed to be aspects of one single category with identical key data names. I need to think about it.

.


Also, the definition of raw here is also up for discussion. Lab XRDs often have strip detectors, so the "raw" data you get off them is already processed to combine the few hundred detectors that exist in "the" detector. And large facilities aren't simpler.

rowlesmr commented:
> `_pd_instr_detector.id` should be linked to `diffrn_detector`

but if we do that in an imgCIF-sense, we get linked back to `_diffrn.id`? Why does changing the specimen temperature affect the detector? or are you trying to describe that data coming from it?

jamesrhester commented:
OK. I've gone back to basics to try and figure this out.

## Axiom: 
 * It must be possible for `PD_MEAS`, `PD_CALC`, and `PD_PROC` to be looped together or separately. This is a fundamental design requirement of the original dictionary (This is the same case as the U_aniso being allowed to loop together or separately from the atomic positions in the core dictionary.) 

The deductions that follow:
    
1.  `PD_MEAS/CALC/PROC` are all child categories of `PD_DATA` (this official "child" relationship of Loop categories is what allows separate/together looping). This means those categories must share key data names, including `_pd_diffractogram.id`. The current dictionary fulfills this requirement.
2.  Therefore `pd_diffractogram` cannot be used exclusively to label raw/processed/calculated data. It is instead equivalent to `pd_data_overall`, providing information about the data as a whole.
3. As multiple detectors can contribute to `PD_MEAS`, `pd_diffractogram` as overall information cannot always be associated with a single detector.
4. So `_pd_diffractogram.detector_id` is ill-defined and should be removed.

## Linking data to detectors and instruments

Propose retaining/adding: 
1. `_pd_instr_detector.instr_id` linking to `pd_instr`
2. `_pd_diffractogram.instr_id` linking to `pd_instr`
3. `_pd_instr_detector.diffrn_detector_id` linking 2D detector elements to post-sample optics.

How this plays out for linking instrument to data:
### 1D data

1. If raw data are presented in `PD_MEAS`: pd_meas.detector_id -> pd_instr_detector -> pd_instr
2. If processed data only are presented in `PD_PROC`: pd_proc -> pd_diffractogram -> pd_instr. If both (1) and (2) are present, then these values must agree. Likewise if only `PD_MEAS` is present both paths must agree (only one of `_pd_meas.detector_id` or `_pd_diffractogram.instr_id` is required).

### 2D data

1. Raw 2D data are described using the usual imgCIF data names.
2. The resulting diffractogram is given in `PD_PROC`.
3. The instrument is identified as per 1D data,  point (2) above.
4. The detectors are identified as for 1D data with the extra pointer to the 2D detector

## Notes

1. The ID22 "area detector behind analyser crystals" is handled by having multiple `pd_instr_detector`s all pointing to the same `diffrn_detector_element` of a monolithic detector. There is not quite enough information to reduce the data as the relative positioning of the analyser crystals needs to be expressed and there are no data names for this. We don't have to figure that out now.
2. If strip detectors like the Mythen produce 2D raw data, that's one route; if they produce 1D raw data (sum pixels axially internally), that's another. I think the point is that `PD_PROC` are the data that ready for fitting or calibratoin, `PD_MEAS` are the data as far back as you can go towards raw data.
3. I've reversed the link direction for the diffrn_detector_element - pd_instr_detector link so that we aren't creating a new 2D element for every separate post-sample optical setup.

Have I missed anything?

rowlesmr commented:
- We can link multiple `_pd_instr_detector.id` to a single `_diffrn_detector.id`, so we can continue to use `_pdi.id` as a channel proxy*.
- We can have multplie 1D detectors contributing to a single `PD_MEAS` diffractogram.
- We can have a 2D detector giving us `PD_PROC` diffractogram directly. The details on how this data reduction is done is delegated to other things to record.

Questions:
- Can we have multiple 2D detectors giving us a single `PD_PROC` diffractogram? - This has to be a yes. Refer to ID22. What does this look like in a CIF file? Does a link between the source 2D data and the `_pd_proc.diffractogram_id` need to made directly, or is the `_diffrn_detector.id` sufficient (diffractogram.id -> instr.id -> instr_detector.id -> dffrn_detector.id)?
- Can we combine 1D and 2D detectors giving us a single `PD_PROC` diffractogram? - This has to be a yes. 1D data can be stored in `PD_MEAS`, 2D data has to be in imgCIF. `_pd_proc.*` data names just appear, and the proc process is described using other data names, and everything is both hunky and dory.

To do:
- add `_pd_instr_detector.diffrn_detector_id`, to allow pdCIF detectors to be linked to physical detectors.
- redefine `DIFFRN_DETECTOR` to have a key data name, `_dd.id`, matching the type in imgCIF. Is this a multiblock thing or a pdCIF thing? I think the latter.
- do the * thing, below

* I propose we make `_pd_meas.channel` (or detector_element or some other word) to explicitly deal with channel, and leave detector_id to handle physical detectors. We redefine `_pdi.id` so that it should be the ID of a physical detector, while noting (and still allowing, but frowning upon) its use as a channel proxy. Make this an Assigned Real Number?

jamesrhester commented:
> Can we have multiple 2D detectors giving us a single PD_PROC diffractogram? - This has to be a yes. Refer to ID22. What does this look like in a CIF file? Does a link between the source 2D data and the _pd_proc.diffractogram_id need to made directly, or is the _diffrn_detector.id sufficient (diffractogram.id -> instr.id -> instr_detector.id -> dffrn_detector.id)?

The diffrn_detector.id is sufficient, as you point out.

> Can we combine 1D and 2D detectors giving us a single PD_PROC diffractogram? - This has to be a yes. 1D data can be stored in PD_MEAS, 2D data has to be in imgCIF. _pd_proc.* data names just appear, and the proc process is described using other data names, and everything is both hunky and dory. Need to ensure _pd_data.point_id values are appropriately assigned.

Yes, that looks right.

> redefine DIFFRN_DETECTOR to have a key data name, _dd.id, matching the type in imgCIF. Is this a multiblock thing or a pdCIF thing? I think the latter.

We are duplicating what is already in imgCIF, so we should probably put it in multiblock as it is more universal than just powder.

> I propose we make _pd_meas.channel (or detector_element or some other word) to explicitly deal with channel, and leave detector_id to handle physical detectors. We redefine _pdi.id so that it should be the ID of a physical detector, while noting (and still allowing, but frowning upon) its use as a channel proxy. Make this an Assigned Real Number?

That sounds reasonable.

rowlesmr commented:
Should _pd_meas.channel be a Real or Integer?

It is designed to refer to a channel or subsection of a detector to which we can apportion an intensity.

jamesrhester commented:
> Should _pd_meas.channel be a Real or Integer?
> 
> It is designed to refer to a channel or subsection of a detector to which we can apportion an intensity. That would imply integer, but I would like it to be as general as possible.

As a rule it is best for an identifier to never contain any other information (because eventually the information content rules it out as an identifier - hkl was used as an identifier, then we got incommensurate structures and multiple measurements. Oops!). So it should be an integer or even opaque code. A `Real` would also imply that the printed value was an approximation.

rowlesmr commented:
I think this is ready to go now, given that https://github.com/COMCIFS/MultiBlock_Dictionary/pull/31 is merged.

What's in here?

- Updated descriptions in `PD_DATA`, `PD_CALC`, `PD_MEAS`, and `PD_PROC` specifying that they are 1D datasets, and that higher dimensional datasets must be reduced before putting them into one of these.
- Added `_pd_meas.channel` to act as a data name to hold channel number-loke information, rather than making `_pd_meas.detector_id`, and `_pd_instr_detector.id` act both as channel and detector identifiers.
- Clarified that `_pd_meas.detector_id`, and `_pd_instr_detector.id` should both be used to refer to detectors, rather than individual channels, although it doesn't outlaw this in order to maintain compatibility with previous behaviour.
- Gave some more info on radiation sources in `PD_INSTR` description.
- Updated distance descriptions in `PD_INSTR` and  `PD_INSTR_DETECTOR`  data names.
- Added `_pd_instr.radiation_id`, linked to `_diffrn_radiation.id`, so the instrument knows which source is plugged in to it.
- Added `_pd_instr_detector.diffrn_detector_id` linked to `_diffrn_detector.id`.
- Added `_pd_instr_detector.diffrn_id` linked to `_diffrn.id`. This is there only to maintain compatibility with imgCIF and `DIFFRN_DETECTOR`. It shouldn't be used; if it were, there would be a `_pd_instr_detector.id` per `_diffrn.id`! If you want a `_diffrn.id` value, look up `_pd_diffractogram.diffrn_id`. Words to this effect are in the dataname description.
- Added `_pd_instr_detector.instr_id` to link the detectors to the instrument (`_pd_instr.id`) they're attached to.

jamesrhester commented on cif_pow.dic:
```diff
@@ -7924,16 +8033,61 @@ save_pd_instr.soller_eq_spec_detc
-    defined. Where a single detector is used, this may be omitted.
+    An arbitrary code which identifies the detector.
+
+    Although it is possible, it should not be used to describe a channel
```
I would delete "Although it is possible" because that is opening a gate for software authors. Perhaps the text of Vol G might discuss legacy use as a channel identifier. I wonder how many datasets are around that actually use this as a channel identifier in a way that is problematic?

jamesrhester reviewed (commented):
Great work. As noted in my review comments, there is some finessing going on regarding a change in the way of expressing raw TOF measurements, but it was always a stretch as Vol G 3.3.8.5 notes. Going forward we'll need to use imgCIF pointing to external images to handle the dataset size.

jamesrhester commented:
I would have made some edits myself but not sure how to make them appear as "suggested edits" rather than new commits to the branch.

rowlesmr commented on cif_pow.dic:
```diff
@@ -4326,6 +4326,13 @@ save_PD_DATA
+    function of 2θ, d-spacing, time-of-flight, position, or other
+    applicable diffraction-related coordinate as defined in PD_MEAS
+    or PD_PROC. Higher dimensional data must first be reduced in some
+    manner before it can be assigned to data names in these categories.
```
done

jamesrhester replied:
Yes, there is a blank semantic space between the pure raw data and the reduced data, for which no data names are really defined. If we wanted to document the stages of data reduction we'd probably make it so that the results of each stage were in a `PD_PROC` loop in a separate data block, but I think this is too big a project to tackle right now. Getting the raw data properly described is most important.

jamesrhester replied:
> I assume that imgCIF is all setup to handle those calibrations and able to deliver ready-to-use data?

That would be a 'no'. It provides all the information you would need, i.e. location in real space of every pixel, but there are no data names to record further data reduction steps. That's up to mmCIF or cif_core or pdCIF.

rowlesmr replied:
Then would it be too much to allow reduced data to fit into MEAS?

rowlesmr replied:
This is worthy of a larger discussion.

rowlesmr commented on cif_pow.dic:
```diff
@@ -5095,14 +5144,17 @@ save_pd_meas.detector_id
+    Although it is possible to use this data name to describe channel
+    number in a position-sensitive, energy-dispersive or other
+    multiple-detector instrument, _pd_meas.channel should alwayes
+    be used in lieu where possible.
```
The OG definition of `_pd_meas_detector_id` was
>A code or number which identifies the measuring detector or
               channel number in a position-sensitive, energy-dispersive
               or other multiple-detector instrument.

I don't know how much legacy is hanging around out there, so I put in `_pd_meas.channel` to explicitly separate channel number from a physical detector, whilst still allowing the description of detector_id to cover channel. That can be tightened to cover only legacy cases, but there isn't any real way to deprecate it, as the OG definition had the data name pulling double duty.

The small amount of TOF that I've seen has had one 1D diffractogram per bank; pretty sure one of the examples in the pdCIFplotter paper is TOF.

rowlesmr replied:
>Currently it seems that I could have multiple channels for a single detector_id, which case both detector id and channel id provide non-identical information that is interpreted differently.

Yes, OG definition had detector.id doing two things.

jamesrhester replied:
Oh well, good to disentangle these now, I guess.

rowlesmr commented on cif_pow.dic:
```diff
@@ -7924,16 +8033,61 @@ save_pd_instr.soller_eq_spec_detc
+
+    Included to maintain compatibility with DIFFRN_DETECTOR and imgCIF.
+    Do not use this value. Use the value associated with the appropriate
+    _pd_diffractogram.diffrn_id in any lookup associated with DIFFRN_DETECTOR.
```
I can grok that. So the idea would be that the "first" `_diffrn.id` is stored there, and only that value is used to define the detectors, so we don't have 3 squillion detectors with all the `_diffrn.id`s that exist.

jamesrhester commented on cif_pow.dic:
```diff
@@ -4782,6 +4795,13 @@ save_PD_MEAS
+    The data contained within PD_MEAS only represent one-dimensional
+    datasets, that is, intensity as some function of 2θ, d-spacing,
+    time-of-flight, position, or other applicable diffraction-related
+    coordinate as defined. Higher dimensional data must first be reduced
```
I don't think `PD_PROC` needs things like time-of-flight at this stage. In my mind we have raw data, we have the final reduced data, and the intermediate steps are not yet described. Filling in that blank is a project for the future.

jamesrhester commented on cif_pow.dic:
```diff
@@ -7924,16 +8033,61 @@ save_pd_instr.soller_eq_spec_detc
+    _description.text
+;
+    The diffrn detector (see _diffrn_detector.id) to which the pdCIF
+    detector belongs.
```
True. "Described using data names in the imgCIF and core dictionaries" would work.
